File: lib/arch_chroot/arch_chroot.py
import subprocess

_root_mount_point = "/mnt"


class ArchChrootExecutionError(Exception):
    def __init__(self, return_code, error_message):
        self.return_code = return_code
        self.error_message = error_message


def run(command: str) -> None:
    try:
        subprocess.run(
            ["arch-chroot", _root_mount_point, command],
            check=True,
            text=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )
    except subprocess.CalledProcessError as e:
        raise ArchChrootExecutionError(e.returncode, e.stderr)

File: pkg/arch_chroot/arch_chroot.go
package arch_chroot

import (
	"io"
	"os/exec"
	"strings"
)

// mountPoint is the mount point of the system to chroot into
const mountPoint string = "/mnt"

// shell is the shell that will be used to execute chroot commands
const shell string = "/bin/bash"

// Executes the command in a shell using arch-chroot.
//
// It executes: arch-chroot [mount_point] [shell] -c [command]
//
// It can return types of errors:
//   - ArchChrootError
func Run(commands ...string) error {
	var command strings.Builder
	for i, cmd := range commands {
		command.WriteString(cmd)

		if i != len(commands)-1 {
			command.WriteString(" && ")
		}
	}

	cmd := exec.Command("arch-chroot", mountPoint, shell, "-c", command.String())
	stderr, err := cmd.StderrPipe()
	if err != nil {
		return ArchChrootError{err: err}
	}

	err = cmd.Run()
	if err != nil {
		stdErrOutput, _ := io.ReadAll(stderr)
		return ArchChrootError{
			command: command.String(),
			stdErr:  string(stdErrOutput),
			err:     err,
		}
	}

	return nil
}
